### DISCUSSION TITLE
Should integrations be allowed to expose new ports on the Home Assistant host?

### DISCUSSION THREAD

**Discussion by jacobstim** [home-assistant/architecture#881] [General] [answered]
## Context
In order to build some integrations, Home Assistant (HA) might need to be discovered and act as a discoverable "gateway" rather than as a client that pulls in data from another source. For example:
-  suppose an integration wants to expose Modbus TCP data as a server (TCP port 502)
-  wants to act as a gateway in a home energy management solution (the case at hand), exposing HA entities as controllable for other devices. 

For example, the "emulated_hue" bridge exposes an SSDP endpoint and a REST endpoint.

This raises the question: **should integrations be allowed to expose additional ports where they listen for incoming calls**? Is this the way forward for increasing the integration capabilities for HA, or what alternatives exist to perform this style of gateway integrations? What are the thoughts and current line of thinking?

## Alternative
As an alternative, we can offload the work to the user and ask them to spin up a "mediator" service that provides the gateway functionality and gets fed through a regular HA integration the data it needs to serve. This isolates the gateway from the HA environment

## Pro/contra
Pro:
- Tight integration with existing HA entities possible
- Easy experience for the user: just install the integration in HA - no need to run additional containers or external services

Con:
- Security risk: opening more ports opens options for more exploits
- Additional maintenance complexity for integration builders (as compared to putting the hosting outside HA)

**Comment by frenck** [comment, marked as answer]
Multiple integration already do so, we have no ruling against it.

../Frenck

**Reply by balloob** [reply]
Just one extra note: it should be avoided if possible. If the other service supports HTTP on non-standard ports, ie sending data via webhooks, the built-in http/webhook integration should be used.
